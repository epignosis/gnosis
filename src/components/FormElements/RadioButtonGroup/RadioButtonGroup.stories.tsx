import React, { useState } from "react";
import { Story } from "@storybook/react";
import RadioButtonGroupComponent, { RadioGroupProps } from "./RadioButtonGroup";

export default {
  title: "components/Form Elements/Radio Button Group",
  component: RadioButtonGroupComponent,
  args: {
    size: "md",
    id: "radioButtonGroupStory",
    className: "radioButtonGroupStory",
    options: [
      {
        label: "All",
        value: "all",
      },
      {
        label: "In progress",
        value: "progress",
      },
      {
        label: "Completed",
        value: "completed",
      },
      {
        label: "Failed",
        value: "failed",
      },
    ],
  },
  argTypes: {
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
  },
};

export const RadioButtonGroup: Story<RadioGroupProps> = (args) => {
  const [value, setValue] = useState("");

  return <RadioButtonGroupComponent id="radio-group" {...args} value={value} onChange={setValue} />;
};
