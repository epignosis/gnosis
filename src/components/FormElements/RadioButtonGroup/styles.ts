import { css, Theme, SerializedStyles } from "@emotion/react";
import { InputSize } from "../Input/Input";

const height = {
  md: "2.5rem",
  lg: "3rem",
};

export const radioButton = (
  { typeScaleSizes, formElements: { radioButtonGroup } }: Theme,
  { size }: { size: InputSize },
): SerializedStyles => {
  const fontSizes = {
    md: typeScaleSizes.sm,
    lg: typeScaleSizes.md,
  };

  return css`
    height: ${height[size]};
    font-size: ${fontSizes[size]};
    padding: 0 1.4rem;
    background: ${radioButtonGroup.normalBackground};
    color: ${radioButtonGroup.normalFontColor};
    border: none;
    border-inline-start: 1px solid ${radioButtonGroup.borderColor};
    cursor: pointer;

    &:first-of-type {
      border-start-start-radius: 5px;
      border-end-start-radius: 5px;
      border-inline-start: none;
    }

    &:last-of-type {
      border-start-end-radius: 5px;
      border-end-end-radius: 5px;
    }

    &.is-selected {
      background: ${radioButtonGroup.selectedBackground};
      color: ${radioButtonGroup.selectedFontColor};
      border-inline-start-color: ${radioButtonGroup.selectedBackground};
    }
  `;
};
