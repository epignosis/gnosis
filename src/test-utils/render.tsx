import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
