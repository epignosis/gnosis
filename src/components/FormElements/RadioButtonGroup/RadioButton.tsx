import React, { FC, ReactNode, MouseEvent } from "react";
import classNames from "classnames";
import { SerializedStyles } from "@emotion/react";
import { InputSize } from "../Input/Input";
import { radioButton } from "./styles";

export type RadioButtonProps = {
  index: number;
  label: ReactNode;
  value: string;
};

export type UiRadioButtonProps = RadioButtonProps & {
  selectedValue: string;
  onClick: (value: string) => void;
  size: InputSize;
};

const RadioButton: FC<UiRadioButtonProps> = ({
  index,
  label,
  value,
  selectedValue,
  onClick,
  size,
}) => {
  const handleClick = (e: MouseEvent): void => {
    e.preventDefault();
    onClick(value);
  };
  const cssClasses = classNames({
    "is-selected": value === selectedValue,
  });

  return (
    <button
      id={`radio-${index}`}
      className={cssClasses}
      onClick={handleClick}
      css={(theme): SerializedStyles => radioButton(theme, { size })}
      aria-selected={value === selectedValue}
    >
      {label}
    </button>
  );
};

export default RadioButton;
