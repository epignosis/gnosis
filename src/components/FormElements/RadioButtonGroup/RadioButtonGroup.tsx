import React, { FC } from "react";
import RadioButton, { RadioButtonProps } from "./RadioButton";

export type RadioGroupOption = Pick<RadioButtonProps, "label" | "value">;

export type RadioGroupProps = {
  options: RadioGroupOption[];
  value: string;
  size?: "md" | "lg";
  id?: string;
  className?: string;
  onChange: (selectedValue: string) => void;
};

const RadioButtonGroup: FC<RadioGroupProps> = (props) => {
  const { options, value = "", size = "md", id, className, onChange } = props;

  return (
    <div id={id} className={className}>
      {options.map((option, index) => (
        <RadioButton
          key={option.value}
          index={index}
          size={size}
          selectedValue={value}
          {...option}
          onClick={onChange}
        />
      ))}
    </div>
  );
};

export default RadioButtonGroup;
