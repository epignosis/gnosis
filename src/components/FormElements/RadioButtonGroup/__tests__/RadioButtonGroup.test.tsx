import React from "react";
import { faker } from "@faker-js/faker";
import RadioButtonGroup from "../RadioButtonGroup";
import { screen, render } from "@test-utils/render";

const OPTIONS = [
  {
    label: faker.company.name(),
    value: faker.company.companySuffix(),
  },
  {
    label: faker.commerce.department(),
    value: faker.commerce.productAdjective(),
  },
  {
    label: faker.git.commitMessage(),
    value: faker.git.commitSha(),
  },
  {
    label: faker.hacker.phrase(),
    value: faker.hacker.abbreviation(),
  },
];

describe("<RadioButtonGroup />", () => {
  it("should render correctly", () => {
    render(
      <RadioButtonGroup
        id={faker.hacker.abbreviation()}
        options={OPTIONS}
        value=""
        onChange={jest.fn()}
      />,
    );

    const options = screen.getAllByRole("button");

    expect(options).toHaveLength(4);
    expect(options[0]).toHaveTextContent(OPTIONS[0].label);
    expect(options[1]).toHaveTextContent(OPTIONS[1].label);
    expect(options[2]).toHaveTextContent(OPTIONS[2].label);
    expect(options[3]).toHaveTextContent(OPTIONS[3].label);
  });

  it("should render with a selected value", () => {
    render(
      <RadioButtonGroup
        id={faker.hacker.abbreviation()}
        options={OPTIONS}
        value={OPTIONS[0].value}
        onChange={jest.fn()}
      />,
    );

    const options = screen.getAllByRole("button");

    expect(options).toHaveLength(4);
    expect(options[0]).toHaveAttribute("aria-selected", "true");
  });

  it("matches snapshot", () => {
    const { container } = render(
      <RadioButtonGroup
        id="radio-group-1"
        className="radio-group"
        options={[{ label: "Test label", value: "testValue" }]}
        value=""
        onChange={jest.fn()}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with selected value", () => {
    const { container } = render(
      <RadioButtonGroup
        id="testRadioButtonGroup"
        options={[{ label: "Test label", value: "testValue" }]}
        value="testValue"
        onChange={jest.fn()}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
