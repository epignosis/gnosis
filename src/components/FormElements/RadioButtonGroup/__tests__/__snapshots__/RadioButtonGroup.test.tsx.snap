// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButtonGroup /> matches snapshot 1`] = `
<div>
  <div
    class="radio-group"
    id="radio-group-1"
  >
    <button
      aria-selected="false"
      class=" css-kcclw-radioButton-radioButton"
      id="radio-0"
    >
      Test label
    </button>
  </div>
</div>
`;

exports[`<RadioButtonGroup /> matches snapshot with selected value 1`] = `
<div>
  <div
    id="testRadioButtonGroup"
  >
    <button
      aria-selected="true"
      class="is-selected css-kcclw-radioButton-radioButton"
      id="radio-0"
    >
      Test label
    </button>
  </div>
</div>
`;
